src/DotNetty.Handlers/Tls/TlsHandler.cs

```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
+                // yet, which means we need to trigger the read to ensure we not encounter any stalls.
+                ctx.Read();
+            }
+        }
+
+        /// <summary>Unwraps inbound SSL records.</summary>
+        void Unwrap(IChannelHandlerContext ctx, IByteBuffer packet, int offset, int length, List<object> output)
+        {
+            //bool notifyClosure = false; // todo: netty/issues/137
+            bool pending = false;
+
+            IByteBuffer decodeOut = null;
+
+            try
+            {
+                this.mediationStream.AcceptBytes(packet.GetIoBuffer(offset, length));
+
+                if (!this.EnsureAuthenticated())
+                {
+                    return;
+                }
+
+                Task<int> currentReadFuture = this.pendingReadFuture;
+
+                if (currentReadFuture != null && !currentReadFuture.IsCompleted)
+                {
+                    return;
+                }
+
+                decodeOut = this.pendingReadBuffer ?? ctx.Allocator.Buffer(length);
+
+                ArraySegment<byte> outlet = decodeOut.GetIoBuffer(decodeOut.WriterIndex, decodeOut.WritableBytes);
+
+                int outletOffset = 0; // offset in outlet that we're through with already
+                int read = 0; // total bytes read out during this unwrapping
+                do
+                {
+                    if (currentReadFuture != null)
+                    {
+                        int readInstant = currentReadFuture.Result;
+
+                        Contract.Assert(readInstant > 0);
+
+                        outletOffset += readInstant;
+                        read += readInstant;
+
+                        IByteBuffer swappedOut;
+                        int remainingInput = this.mediationStream.SourceLength; // bytes not yet read by SslStream
+                        if (remainingInput == 0)
+                        {
+                            // SslStream ingested whole input so it might require more input on next ReadAsync call.
+
+                            // regardless of buffer's write capacity - we need to swap it out (for different reasons though)
+                            swappedOut = decodeOut;
+
+                            if (outletOffset < outlet.Count)
+                            {
+                                // current output buffer isn't full - SslStream will likely block =>
+                                // release current result and switch to buffer of FallbackReadBufferSize
+                                decodeOut = ctx.Allocator.Buffer(FallbackReadBufferSize);
+                            }
+                            else
+                            {
+                                // current output buffer is full - very likely there's more data in SslStream's internal state
+                                decodeOut = ctx.Allocator.Buffer(Math.Max(FallbackReadBufferSize, length - read));
```

Comment:
what's the heuristic here? technically we've exhausted all the remaining input? how much do we expect this to be?